File: 模板/python/EulerTour.py
def EulerTour(n, edges, root=0):
    L = [-1] * n
    R = [0] * n
    stack = [(root, 1), (root, 0)]
    ind = 0
    while stack:
        pos, t = stack.pop()
        if t == 0:
            L[pos] = ind
            ind += 1
            for npos in edges[pos]:
                if L[npos] != -1:
                    continue
                stack.append((npos, 1))
                stack.append((npos, 0))
        else:
            R[pos] = ind
 
    return L, R

File: 模板/cpp/EulerTour.cpp
pair<vector<int>, vector<int>> EulerTour(vector<vector<int>>& E, int root = 0) {
    int n = E.size();
    vector<int> IN(n, -1);
    vector<int> OUT(n, 0);
    vector<pair<int, int>> stack = {{root, 1}, {root, 0}};
    int ind = 0;
    while (stack.size()) {
        auto [pos, t] = stack.back();
        stack.pop_back();
        if (!t) {
            IN[pos] = ind++;
            for (auto& npos : E[pos]) if (IN[npos] == -1) {
                stack.emplace_back(pair{npos, 1});
                stack.emplace_back(pair{npos, 0});
            }
        } else OUT[pos] = ind;
    }
    return {IN, OUT};
}